python/ray/rllib/bc/tfpolicy.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import ray
+import tensorflow as tf
+from ray.rllib.a3c.policy import Policy
+
+
+class TFPolicy(Policy):
+    """The policy base class."""
+
+    def __init__(self, registry, ob_space, action_space, config,
+                 name="local", summarize=True):
+        self.registry = registry
+        self.local_steps = 0
+        self.config = config
+        self.summarize = summarize
+        worker_device = "/job:localhost/replica:0/task:0/cpu:0"
+        self.g = tf.Graph()
+        with self.g.as_default(), tf.device(worker_device):
+            with tf.variable_scope(name):
+                self._setup_graph(ob_space, action_space)
+            print("Setting up loss")
+            self.setup_loss(action_space)
+            self.setup_gradients()
+            self.initialize()
+
+    def _setup_graph(self, ob_space, ac_space):
+        raise NotImplementedError
+
+    def setup_loss(self, action_space):
+        raise NotImplementedError
+
+    def setup_gradients(self):
+        grads = tf.gradients(self.loss, self.var_list)
+        self.grads, _ = tf.clip_by_global_norm(grads, self.config["grad_clip"])
+        grads_and_vars = list(zip(self.grads, self.var_list))
+        opt = tf.train.AdamOptimizer(self.config["lr"])
+        self._apply_gradients = opt.apply_gradients(grads_and_vars)
+
+    def initialize(self):
+        if self.summarize:
+            bs = tf.to_float(tf.shape(self.x)[0])
+            tf.summary.scalar("model/policy_loss", self.pi_loss / bs)
+            tf.summary.scalar("model/grad_gnorm", tf.global_norm(self.grads))
+            tf.summary.scalar("model/var_gnorm", tf.global_norm(self.var_list))
+            self.summary_op = tf.summary.merge_all()
+
+        # TODO(rliaw): Can consider exposing these parameters
+        self.sess = tf.Session(graph=self.g, config=tf.ConfigProto(
+            intra_op_parallelism_threads=1, inter_op_parallelism_threads=2,
+            gpu_options=tf.GPUOptions(allow_growth=True)))
```

Comment:
is this the only change in this file compared to `a3c/tfpolicy`? If so, we should just expose these parameters via the default config in A3C and avoid repeating too much code.

Reply:
the bigger change is removing the `assert` in `__init__`; and I didn't want to create a dependency of BC on A3C; we should refactor the policy to be external to either of these

Reply:
I see; there is still a dependency on A3C via `a3c/policy.py`. I guess it would be reasonable to duplicate all of the policy code for now, @ericl?

Reply:
Duplicating sounds good, there's not a lot.

Btw, shouldn't we combine Policy and TFPolicy into one class? There doesn't seem to be much value in separating them.

Reply:
Yeah, probably should combine `bc/policy.py` and `bc/tfpolicy.py`. We can leave the `a3c/policy.py` dependency as it's just an interface.